### PR TITLE
add documentation for cwrap plugins

### PR DESCRIPTION
**Pull request by killeent**
Make it easy to write new plugins and understand how the existing plugins work.

### CODE REVIEW

File: tools/cwrap/cwrap.py [hunk 1]
```diff
@@ -35,11 +35,11 @@ class cwrap(object):
     DEFAULT_PLUGIN_CLASSES = [ArgcountChecker, ConstantArguments, OptionalArguments,
                               ArgumentReferences, BeforeAfterCall, ReturnArguments, GILRelease]
 
-    def __init__(self, source, destination=None, plugins=[], default_plugins=True):
+    def __init__(self, source, destination=None, plugins=None, default_plugins=True):
```
Comment (killeent) [inline tools/cwrap/cwrap.py]: http://python-guide-pt-br.readthedocs.io/en/latest/writing/gotchas/#mutable-default-arguments

**Review by soumith — APPROVED** [review]

File: tools/cwrap/plugins/__init__.py [hunk 2]
```diff
@@ -1,55 +1,418 @@
...
+            type or specifier.
+
+        """
         pass
 
     def get_wrapper_template(self, declaration):
+        """Used to create a code template to wrap the options.
+
+        This function returns a Template string that contains the function call
+        for the overall declaration, including the method definition and { ...
```
Comment (apaszke) [inline tools/cwrap/plugins/__init__.py]: Do you want to complete this before merging?

File: tools/cwrap/plugins/__init__.py [hunk 3]
```diff
@@ -1,55 +1,418 @@
...
+        a constant, or self. The base cwrap plugin has a fallback arg accessor
+        for loading elements from the args PyObject * tuple passed to the
+        function.
+
+        Its best to look at some of the existing Plugins to get a sense of what
+        one might do.
+
+        Args:
+            arg: a dictionary specifying attributes of the arg to be accessed
+            option: ... ??
```
Comment (apaszke) [inline tools/cwrap/plugins/__init__.py]: as everywhere

File: tools/cwrap/plugins/__init__.py [hunk 4]
```diff
@@ -1,55 +1,418 @@
...
+            option.
+
+        Returns:
+            The same code, modified as the plugin sees fit.
+
+        """
         return code
 
     def process_option_code(self, code, option):
+        """Used to modify the entire code body for an option.
```
Comment (apaszke) [inline tools/cwrap/plugins/__init__.py]: It would be helpful to say that body includes the option checks (the whole `else if (...)`).

Comment (killeent) [inline tools/cwrap/plugins/__init__.py]: Done

**Review by apaszke — APPROVED** [review]

**Pull request merged**